src/subscript.c

```diff
@@ -0,0 +1,254 @@
+
+  if (*err) {
+    UNPROTECT(2);
+    return R_NilValue;
+  }
+
+  // Coerce unspecified vectors to integer only if logical indices are
+  // not allowed
+  if (opts->logical == SUBSCRIPT_TYPE_ACTION_ERROR && vec_is_unspecified(subscript)) {
+    struct vctrs_arg* arg = opts->subscript_arg;
+    if (opts->numeric == SUBSCRIPT_TYPE_ACTION_CAST) {
+      subscript = vec_cast(subscript, vctrs_shared_empty_int, arg, NULL);
+    } else {
+      subscript = vec_cast(subscript, vctrs_shared_empty_chr, arg, NULL);
+    }
+  }
+  REPROTECT(subscript, subscript_pi);
+
+  enum subscript_type_action action = SUBSCRIPT_TYPE_ACTION_ERROR;
+  switch (TYPEOF(subscript)) {
+  case LGLSXP: action = opts->logical; break;
+  case INTSXP: action = opts->numeric; break;
+  case STRSXP: action = opts->character; break;
+  default: break;
+  }
+
+  if (action == SUBSCRIPT_TYPE_ACTION_ERROR) {
+    *err = new_error_subscript_type(subscript, opts, R_NilValue, R_NilValue);
+    UNPROTECT(2);
+    return R_NilValue;
+  }
+
+  // FIXME: Handle names in cast methods
+  subscript = r_maybe_duplicate(subscript);
+  REPROTECT(subscript, subscript_pi);
+  r_poke_names(subscript, orig_names);
+
+  UNPROTECT(2);
+  return subscript;
+}
+
+static SEXP obj_cast_subscript(SEXP subscript,
+                               struct vec_as_subscript_opts* opts,
+                               ERR* err) {
+  int dir = 0;
+  struct vctrs_arg* arg = opts->subscript_arg;
+
+  // FIXME: No need for is-coercible check once vec-cast is restricted
+  // to coercible casts
+
+  if (vec_is_coercible(subscript, vctrs_shared_empty_lgl, arg, NULL, &dir)) {
+    return vec_cast(subscript, vctrs_shared_empty_lgl, arg, NULL);
+  }
+
+  if (vec_is_coercible(subscript, vctrs_shared_empty_int, arg, NULL, &dir)) {
+    return vec_cast(subscript, vctrs_shared_empty_int, arg, NULL);
+  }
+
+  if (vec_is_coercible(subscript, vctrs_shared_empty_chr, arg, NULL, &dir)) {
+    return vec_cast(subscript, vctrs_shared_empty_chr, arg, NULL);
+  }
+
+  *err = new_error_subscript_type(subscript, opts, R_NilValue, R_NilValue);
+  return R_NilValue;
+}
+
+static SEXP dbl_cast_subscript_body = NULL;
+static SEXP dbl_cast_subscript(SEXP subscript,
+                               struct vec_as_subscript_opts* opts,
+                               ERR* err) {
+  SEXP out = PROTECT(vec_coercible_cast_e(subscript,
+                                          vctrs_shared_empty_int,
+                                          opts->subscript_arg,
+                                          NULL,
+                                          err));
+  if (*err) {
+    SEXP err_obj = PROTECT(*err);
+    *err = new_error_subscript_type(subscript,
+                                    opts,
+                                    dbl_cast_subscript_body,
+                                    err_obj);
+    UNPROTECT(2);
+    return R_NilValue;
+  }
+
+  // FIXME: Handle names in cast methods
+  SEXP nms = PROTECT(r_names(subscript));
```

Comment:
Names are already handled with `orig_names` in `vec_as_subscript_opts()` right? Also, is it possible this could overwrite `subscript` directly if `vec_coercible_cast_e()`'s call to `vec_cast()` doesn't duplicate?

Reply:
oops I forgot to check this could be removed. I agree it would have been safer to maybe-duplicate, in case the method starts to return cached objects for opitimisation for instance.